File: leetcode/search/410. Split Array Largest Sum/js/main.js
// Time O(N * Log(K)) K - сумма элементов в nums
// Space O(1)
const splitArray = (nums, m) => {
  let left = 0;
  let right = 0;

  for (let x of nums) {
    right += x;
    left = Math.max(left, x);
  }

  let ans = right;

  while (left <= right) {
    let mid = left + Math.floor((right - left) / 2);

    if (calc(mid) > m) {
      left = mid + 1;
    } else {
      right = mid - 1;
      ans = Math.min(ans, mid);
    }
  }

  return ans;

  function calc(mid) {
    let sum = 0;
    let cnt = 1;

    for (let i = 0; i < nums.length; i++) {
      if (sum + nums[i] <= mid) {
        sum += nums[i];
      } else {
        sum = nums[i];
        cnt++;
      }
    }

    return cnt;
  }
};

const splitArray_II = (nums, m) => {
  let n = nums.length;
  let ans = Number.MAX_VALUE;

  dfs(0, 0, 0, 0);

  return ans;

  /*

    Получения всех вариантов разбиения массива на M подмассивов 

    Алгоритм:

    Для каждого элемента в массиве мы можем 
    
    1) Добавить его к предыдущему подмассиву 
    2) Запустить новый подмассив, начинающийся с этого элемента (если число подмассивов не превышает m). 
    
    Сумма текущего подмассива может быть обновлена ​​одновременно.

  */

  function dfs(i, cntSubarrays, curSum, curMax) {
    if (i == n && cntSubarrays == m) {
      ans = Math.min(ans, curMax);
      return;
    }

    if (i == n) {
      return;
    }

    if (i > 0) {
      dfs(i + 1, cntSubarrays, curSum + nums[i], Math.max(curMax, curSum + nums[i]));
    }

    if (cntSubarrays < m) {
      dfs(i + 1, cntSubarrays + 1, nums[i], Math.max(curSum, nums[i]));
    }
  }
};

File: leetcode/search/410. Split Array Largest Sum/py/main.py
class Solution:
    # Time O(N * Log(sumofarray))
    # Space O(N)
    def splitArray(self, nums, m):
        if m == len(nums):
            return max(nums)

        lo = 1
        hi = sum(nums)
    
        while lo < hi:
            maxSubarraySum = lo + ((hi - lo ) // 2)
            
            cntSubArrays = self.getCntSubarrays(nums, maxSubarraySum)

            # maxSubarraySum is small, need increase it
            if cntSubArrays > m:
                lo = maxSubarraySum + 1
            else:
                hi = maxSubarraySum
                
        return lo
    
    def getCntSubarrays(self, nums, maxSum):
        currSum = 0
        cnt = 1

        for i in range(len(nums)):
            # maxSum is small
            if nums[i] > maxSum:
                return float('inf')

            if nums[i] + currSum <= maxSum:
                currSum += nums[i]
            else:
                cnt += 1
                currSum = nums[i]
        return cnt

    def splitArray_II(self, nums, m):
        if len(nums) == m:
            return max(nums)

        N = len(nums)
        prefix_sum = [0] * (N + 1)
        inf = float('inf')
        dp = [[-1] * (m + 1) for _ in range(N)]

        for i in range(1, N + 1):
            prefix_sum[i] = prefix_sum[i - 1] + nums[i - 1]
        
        def dfs(index, k):        
            if k < 0:
                return 0
            
            if k == 1:
                return prefix_sum[N] - prefix_sum[index]

            if dp[index][k] != -1:
                return dp[index][k]

            res = inf
            
            for i in range(index, N - 1):
                val = prefix_sum[i + 1] - prefix_sum[index]
                res = min(res, max(val, dfs(i + 1, k - 1)))

            dp[index][k] = res

            return res

        return dfs(0, m)
